File: scala/forth/src/main/scala/Forth.scala
import ForthError.ForthError
import scala.util.Try
import Forth._


case class ForthState(stack: List[Int] = List(),
    newWords:Map[WordName, WordDefinition] = Map()) extends ForthEvaluatorState
{
  def push(elem: Int): ForthOpResult[Unit] =
    Right(((), copy(stack = elem::stack)))

  // FP style: return value and new state
  def pop: ForthOpResult[Int] =
    if (stack.isEmpty) Left(ForthError.StackUnderflow)
    else Right((stack.head, copy(stack = stack.tail)))

  def addNewWord(wordName: WordName, wordDefinition: WordDefinition): ForthOpResult[Unit] = {
    val newWord = (wordName, wordDefinition)
    Right(((), copy(newWords = newWords + newWord)))
  }

  override def toString = stack.reverse mkString " "
}

class Forth extends ForthEvaluator {
  override def eval(input: String): Either[ForthError, ForthEvaluatorState] =
    evalForth(parseInput(input), EmptyEvaluationState)
}

object Forth {
  type EvaluationState = Either[ForthError, ForthState]
  implicit class EvaluationStateOps(self: EvaluationState) {
    def >>=[A](f: ForthResult): EvaluationState = self.right.flatMap(f)
  }
  val EmptyEvaluationState: EvaluationState = Right(ForthState())

  // FP style: return value and new state
  type ForthOpResult[+A] = Either[ForthError, (A, ForthState)]
  type ForthOp[+A] = ForthState => ForthOpResult[A]
  type ForthResult = ForthState => EvaluationState
  implicit def forthOpToForthResult(forthOp: ForthOp[_]): ForthResult =
    fs =>  forthOp(fs).right.map { case (_, state) => state }

  type ParsedInput = List[String]
  type WordName = String
  type WordDefinition = List[String]

  def evalForth(parsedInput: ParsedInput, state: EvaluationState): EvaluationState =
    parsedInput match {
      case Nil => state
      case ":"::wordName::xs => {
        if (parseNumber(wordName).isDefined) Left(ForthError.InvalidWord)
        else {
          val (newWordDef, _::parsedInputRest) = xs span (_ != ";")
          val nextState = state >>= doAddNewWord(wordName.toUpperCase, newWordDef)
          evalForth(parsedInputRest, nextState)
        }
      }
      case x::xs => {
        val maybeWordDef: Option[WordDefinition] =
          state.fold(_ => None, _.newWords.get(x.toUpperCase))
        val nextParsedInput = maybeWordDef.map (_ ++ xs) getOrElse xs
        val maybeNumber = parseNumber(x)
        val nextState: EvaluationState =
          if (maybeNumber.isDefined) state >>= doPush(maybeNumber.get)
          else if (maybeWordDef.isDefined) state
          else x.toUpperCase match {
            case "+" => state >>= doBinaryOp(_ + _)
            case "-" => state >>= doBinaryOp(_ - _)
            case "*" => state >>= doBinaryOp(_ * _)
            case "/" => state >>= doDiv
            case "DUP" => state >>= doDup
            case "DROP" => state >>= doDrop
            case "SWAP" => state >>= doSwap
            case "OVER" => state >>= doOver
            case _ => Left(ForthError.InvalidWord)
          }

        evalForth(nextParsedInput, nextState)
      }
    }

  def parseInput(input: String): ParsedInput = {
    if (input.trim.isEmpty) List()
    else input.split("""[\W&&[^+-/*:;]]""").filter(_.nonEmpty) toList
  }

  def parseNumber(str: String): Option[Int] =
    Try(str.toInt).toOption

  def map[A,B](s: ForthOp[A])(f: A => B): ForthOp[B] =
    fs =>
      s(fs).right.map {
        case (a, fs1) => (f(a), fs1)
      }

  // for-comprehensions w/ Either have a bug when matching tuples:
  // https://issues.scala-lang.org/browse/SI-5589
  // so I had to go for flatMap instead
  def flatMap[A,B](f: ForthOp[A])(g: A => ForthOp[B]): ForthOp[B] =
    fs =>
      f(fs).right.flatMap {
        case (a, fs1) => g(a)(fs1)
      }

  def doPush(elem: Int): ForthOp[Unit] = _.push(elem)
  def doPop: ForthOp[Int] = _.pop
  def doAddNewWord(wn: WordName, wd: WordDefinition): ForthOp[Unit] = _.addNewWord(wn, wd)

  def doBinaryOp(op: (Int,Int) => Int): ForthResult =
    flatMap(doPop) { x =>
      flatMap(doPop) { y =>
        doPush(op(y, x))
      }
    }

  def doDiv: ForthResult =
    flatMap(doPop) { y =>
      flatMap(doPop) { z =>
        if (y != 0) doPush(z / y)
        else _ => Left(ForthError.DivisionByZero)
      }
    }

  def doDup: ForthResult =
    flatMap(doPop) { y =>
      flatMap(doPush(y)) { _ =>
        doPush(y)
      }
    }

  def doDrop: ForthResult = doPop

  def doSwap: ForthResult =
    flatMap(doPop) { x =>
      flatMap(doPop) { y =>
        flatMap(doPush(x)) { _ =>
          doPush(y)
        }
      }
    }

  def doOver: ForthResult =
    flatMap(doPop) { x =>
      flatMap(doPop) { y =>
        flatMap(doPush(y)) { _ =>
          flatMap(doPush(x)) { _ =>
            doPush(y)
          }
        }
      }
    }
}

File: haskell/forth/Forth.hs
{-# LANGUAGE OverloadedStrings #-}
module Forth
  ( ForthError(..)
  , ForthState
  , evalText
  , formatStack
  , empty
  ) where

import Data.Text (Text, split, pack, unpack, toUpper)
import Data.Maybe (isJust, fromJust)
import Data.Char (isAlphaNum, isSymbol)

type ParsedInput = [String]
type ForthResult = Either ForthError ForthState

type WordName = String
type WordDefinition = [String]

data ForthState = ForthState { stack::[Int], newWords::[(WordName, WordDefinition)] }

empty :: ForthState
empty = ForthState [] []

push :: Int -> ForthState -> ForthState
push x (ForthState xs nws) = ForthState (x:xs) nws

pop :: ForthState -> Either ForthError (Int, ForthState)
pop (ForthState (x:xs) nws) = Right (x, ForthState xs nws)
pop _ = Left StackUnderflow

lookupWordDefinition :: WordName -> ForthState -> Maybe WordDefinition
lookupWordDefinition w = lookup w . newWords

data ForthError
     = DivisionByZero
     | StackUnderflow
     | InvalidWord
     | UnknownWord Text
     deriving (Show, Eq)

-- | Evaluate an input Text, returning the new state
evalText :: Text -> ForthState -> ForthResult
evalText t s = evalForth (parseInput t) (Right s)

-- foldl over ParsedInput would be more natural,
-- but I had no better idea to account for new word definitions
evalForth :: ParsedInput -> ForthResult -> ForthResult
evalForth [] s = s
evalForth (":":wordName:xs) forthResult =
  evalForth parsedInputRest newForthResult
  where
    newForthResult = do
      (ForthState s nws) <- forthResult
      wd <- newWordDef
      return (ForthState s (wd:nws))

    newWordDef :: Either ForthError (WordName, WordDefinition)
    newWordDef = if isJust $ parseNumber wordName
      then Left InvalidWord
      else Right (upperCaseWordName, wordDef)

    upperCaseWordName = toUpperCase wordName
    (wordDef, parsedInputRest) = getWordDef [] xs

    getWordDef :: [String] -> ParsedInput -> (WordDefinition, ParsedInput)
    getWordDef wd (";":ys) = (reverse wd, ys)
    getWordDef wd (y:ys) = getWordDef (y:wd) ys
evalForth (x:xs) s = evalForth nextParsedInput nextForthResult
  where
    nextForthResult
      | isJust maybeWordDef = s
      | x == "+" = s >>= doBinaryOp (+)
      | x == "-" = s >>= doBinaryOp (-)
      | x == "*" = s >>= doBinaryOp (*)
      | x == "/" = s >>= doDiv
      | upperCaseWord == "DUP" = s >>= doDup
      | upperCaseWord == "DROP" = s >>= doDrop
      | upperCaseWord == "SWAP" = s >>= doSwap
      | upperCaseWord == "OVER" = s >>= doOver
      | isJust maybeNumber = s >>= doPushNumber (fromJust maybeNumber)
      | otherwise = Left $ UnknownWord $ pack x
    nextParsedInput = maybe xs (++ xs) maybeWordDef
    maybeWordDef = either (const Nothing) (lookupWordDefinition upperCaseWord) s
    maybeNumber = parseNumber x
    upperCaseWord = toUpperCase x

doBinaryOp :: (Int -> Int -> Int) -> ForthState -> ForthResult
doBinaryOp op s = do
  (y, s1) <- pop s
  (z, s2) <- pop s1
  Right (push (z `op` y) s2)

doDiv :: ForthState -> ForthResult
doDiv s = do
  (y, s1) <- pop s
  (z, s2) <- pop s1
  if y /= 0 then Right (push (z `div` y) s2)
            else Left DivisionByZero

doDup :: ForthState -> ForthResult
doDup s = do
  (y, s1) <- pop s
  s2 <- Right (push y s1)
  Right (push y s2)

doDrop :: ForthState -> ForthResult
doDrop s = do
  (_, s1) <- pop s
  Right s1

doSwap :: ForthState -> ForthResult
doSwap s = do
  (y, s1) <- pop s
  (z, s2) <- pop s1
  s3 <- Right (push y s2)
  Right (push z s3)

doOver :: ForthState -> ForthResult
doOver s = do
  (y, s1) <- pop s
  (z, s2) <- pop s1
  s3 <- Right (push z s2)
  s4 <- Right (push y s3)
  Right (push z s4)

doPushNumber :: Int -> ForthState -> ForthResult
doPushNumber n s = Right (push n s)

toUpperCase :: String -> String
toUpperCase = unpack . toUpper . pack

parseNumber :: String -> Maybe Int
parseNumber str = case reads str :: [(Int,String)] of [(n,"")] -> Just n
                                                      _ -> Nothing

parseInput :: Text -> ParsedInput
parseInput = map unpack . split notWordChar
  where
-- did not want to dive into regex matching, so I tried to keep it simple
    notWordChar c = not (isAlphaNum c || isSymbol c || c `elem` "+-/*:;")

-- | Return the current stack as Text with the element
-- on top of the stack being the rightmost element in the output
formatStack :: ForthState -> Text
formatStack = pack . unwords . map show . reverse . stack
